scalding-core/src/test/scala/com/twitter/scalding/ExecutionOptimizationRulesTest.scala

```diff
@@ -0,0 +1,267 @@
+    pipe.map(_.writeThrough(new MemorySource[Int]()))
+
+  val mappedOrFlatMapped =
+    Gen.oneOf(mapped(pipe), flatMapped(pipe))
+
+  val zippedWrites =
+    zipped(write(TypedPipeGen.genWithIterableSources), write(TypedPipeGen.genWithIterableSources))
+
+  val mappedWrites =
+    mapped(write(TypedPipeGen.genWithIterableSources))
+
+  val zippedFlatMapped =
+    Gen.oneOf(
+      zipped(flatMapped(pipe), flatMapped(pipe)),
+      zipped(mappedWrites, flatMapped(pipe)),
+      zipped(flatMapped(pipe), mappedWrites)
+    )
+
+  val zippedMapped =
+    Gen.oneOf(
+      zipped(mappedWrites, mappedOrFlatMapped),
+      zipped(mappedOrFlatMapped, mappedWrites)
+    )
+
+  val genExec =
+    Gen.oneOf(
+      zippedWrites,
+      zipped(mappedOrFlatMapped, write(TypedPipeGen.genWithIterableSources)),
+      zipped(write(TypedPipeGen.genWithIterableSources), mappedOrFlatMapped)
+    )
+
+  val iterableExec =
+    Gen.oneOf(
+      zippedWrites,
+      zippedFlatMapped,
+      zippedMapped,
+      zipped(mappedOrFlatMapped, write(TypedPipeGen.genWithIterableSources)),
+      zipped(write(TypedPipeGen.genWithIterableSources), mappedOrFlatMapped)
+    ).map { exec =>
+      exec flatMap {
+        case (left, right) => left.toIterableExecution.zip(right.toIterableExecution)
+      } map {
+        case (left, right) => left ++ right
+      } map {
+        _.toList.sorted
+      }
+    }
+
+  import ExecutionOptimizationRules._
+
+  val allRules = List(
+    ZipWrite,
+    ZipMap,
+    ZipFlatMap,
+    MapWrite
+  )
+
+  def genRuleFrom(rs: List[Rule[Execution]]): Gen[Rule[Execution]] =
+    for {
+      c <- Gen.choose(1, rs.size)
+      rs <- Gen.pick(c, rs)
+    } yield rs.reduce(_.orElse(_))
+
+  val genRule = genRuleFrom(allRules)
+
+  def invert[T](exec: Execution[T]) =
+    assert(toLiteral(exec).evaluate == exec)
+
+  test("randomly generated executions trees are invertible") {
+    forAll(genExec) { exec =>
+      invert(exec)
+    }
+  }
+
+  test("optimization rules are reproducible") {
+    implicit val generatorDrivenConfig: PropertyCheckConfiguration = PropertyCheckConfiguration(minSuccessful = 500)
+
+    forAll(genExec, genRule) { (exec, rule) =>
+      val optimized = ExecutionOptimizationRules.apply(exec, rule)
+      val optimized2 = ExecutionOptimizationRules.apply(exec, rule)
+      assert(optimized == optimized2)
+    }
+  }
+
+  test("standard rules are reproducible") {
+    implicit val generatorDrivenConfig: PropertyCheckConfiguration = PropertyCheckConfiguration(minSuccessful = 500)
+
+    forAll(genExec) { exec =>
+      val optimized = ExecutionOptimizationRules.stdOptimizations(exec)
+      val optimized2 = ExecutionOptimizationRules.stdOptimizations(exec)
+      assert(optimized == optimized2)
+    }
+  }
+
+  def runAndCompare[A](left: Execution[A], right: Execution[A]) = {
+    val config = Config.unitTestDefault
+
+    assert(
+      left.waitFor(config, Local(true)).get ==
```

Comment:
isn't this test still applying the optimization rules inside of `def run`? I think we need a config flag to disable running optimizations, similar to TypedPipe case, and if the Config has that, do no optimization. Then in this test, we run the optimizations externally and do none inside.

What do you think?

Reply:
yep, nice.